web/src/main/java/it/pagopa/selfcare/dashboard/web/security/ExchangeTokenServiceV2.java

```diff
@@ -0,0 +1,383 @@
+        Assert.notNull(selcClaims, "Session token claims is required");
+        TokenExchangeClaims claims = new TokenExchangeClaims(selcClaims);
+        claims.setId(UUID.randomUUID().toString());
+        claims.setIssuer(issuer);
+        User user = userService.getUserByInternalId(UUID.fromString(userId));
+
+        String email = Optional.ofNullable(user.getWorkContact(institution.getId()))
+                .map(workContract -> Objects.nonNull(workContract.getEmail())
+                        ? workContract.getEmail().getValue()
+                        : ""
+                )
+                .orElse(null);
+
+        claims.setEmail(email);
+        claims.setInstitution(institution);
+        claims.setDesiredExpiration(claims.getExpiration());
+        claims.setIssuedAt(new Date());
+        claims.setExpiration(Date.from(claims.getIssuedAt().toInstant().plus(duration)));
+
+        return claims;
+    }
+
+
+    private Map<String, ProductGrantedAuthority> retrieveProductsFromInstitutionAndUser(String institutionId, String userId) {
+        return this.userApiConnector.getProducts(institutionId, userId)
+                .getProducts()
+                .stream()
+                .collect(Collectors.toMap(
+                        it.pagopa.selfcare.dashboard.connector.model.user.OnboardedProduct::getProductId,
+                        product -> new ProductGrantedAuthority(product.getRole(), List.of(product.getProductRole()), product.getProductId())));
+    }
+
+
+    private String createJwts(TokenExchangeClaims claims) {
+        return Jwts.builder()
+                .setClaims(claims)
+                .signWith(SignatureAlgorithm.RS256, jwtSigningKey)
+                .setHeaderParam(JwsHeader.KEY_ID, kid)
+                .setHeaderParam(Header.TYPE, Header.JWT_TYPE)
+                .compact();
+    }
+
+    private void retrieveAndSetGroups(Institution institution, String institutionId, String productId, String userId) {
+        Page<UserGroupInfo> groupInfos = groupService.getUserGroups(Optional.of(institutionId),
+                Optional.ofNullable(productId),
+                Optional.of(UUID.fromString(userId)),
+                Pageable.ofSize(100));// 100 is a reasonably safe number to retrieve all groups related to a generic user
+        if (groupInfos.hasNext()) {
+            log.warn(String.format("Current user (%s) is member of more than 100 groups related to institution %s and product %s. The Identity Token will contain only the first 100 records",
+                    userId,
+                    institutionId,
+                    productId));
+        }
+        if (!groupInfos.isEmpty()) {
+            institution.setGroups(groupInfos.stream()
+                    .map(UserGroupInfo::getId)
+                    .toList());
+        }
+    }
+
+    private Institution retrieveInstitution(String institutionId, List<ProductGrantedAuthority> productGrantedAuthorities, boolean isBillingToken) {
```

Comment:
Could we move this logic into a separate service (adding also a mapper) avoding to have private methods into this service?